qiskit/primitives/base_estimator.py

```diff
@@ -0,0 +1,236 @@
+
+* parameter values (:math:`\theta_k`): list of sets of values
+  to be bound to the parameters of the quantum circuits.
+  (list of list of float)
+
+The output is the list of expectation values.
+
+.. math::
+
+    \Braket{\psi_i(\theta_k)|H_j|\psi_i(\theta_k)}
+
+
+The estimator object is expected to be close()d after use or
+accessed inside "with" context
+and the objects are called with parameter values and run options
+(e.g., ``shots`` or number of shots).
+
+Here is an example of how estimator is used.
+
+.. code-block:: python
+
+    from qiskit.circuit.library import RealAmplitudes
+    from qiskit.quantum_info import SparsePauliOp
+
+    psi1 = RealAmplitudes(num_qubits=2, reps=2)
+    psi2 = RealAmplitudes(num_qubits=2, reps=3)
+
+    params1 = psi1.parameters
+    params2 = psi2.parameters
+
+    H1 = SparsePauliOp.from_list([("II", 1), ("IZ", 2), ("XI", 3)])
+    H2 = SparsePauliOp.from_list([("IZ", 1)])
+    H3 = SparsePauliOp.from_list([("ZI", 1), ("ZZ", 1)])
+
+    with Estimator([psi1, psi2], [H1, H2, H3], [params1, params2] as e:
+        theta1 = [0, 1, 1, 2, 3, 5]
+        theta2 = [0, 1, 1, 2, 3, 5, 8, 13]
+        theta3 = [1, 2, 3, 4, 5, 6]
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)> ]
+        psi1_H1_result = e([theta1], grouping=[(0, 0)])
+        print(psi1_H1_result)
+
+        # calculate [ <psi1(theta1)|H2|psi1(theta1)>, <psi1(theta1)|H3|psi1(theta1)> ]
+        psi1_H23_result = e([theta1]*2, grouping=[(0, 1), (0, 2)])
+        print(psi1_H23_result)
+
+        # calculate [ <psi2(theta2)|H2|psi2(theta2)> ]
+        psi2_H2_result = e([theta2], grouping=[(1, 1)])
+        print(psi2_H2_result)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>, <psi1(theta3)|H1|psi1(theta3)> ]
+        psi1_H1_result2 = e([theta1, theta3], grouping=[(0, 0), (0, 0)])
+        print(psi1_H1_result2)
+
+        # calculate [ <psi1(theta1)|H1|psi1(theta1)>,
+        #               psi2(theta2)|H2|psi2(theta2)>,
+        #               <psi1(theta3)|H3|psi1(theta3)> ]
+        psi12_H23_result = e([theta1, theta2, theta3], grouping=[(0, 0), (1, 1), (0, 2)])
+        print(psi12_H23_result)
+
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Callable, Iterable, Sequence
+
+from qiskit.circuit import QuantumCircuit, Parameter
+from qiskit.circuit.parametertable import ParameterView
+from qiskit.exceptions import QiskitError
+from qiskit.quantum_info import SparsePauliOp
+
+from .estimator_result import EstimatorResult
+
+
+@dataclass(frozen=True)
+class Group:
+    """The dataclass represents indices of circuit and observable."""
+
+    circuit_index: int
+    observable_index: int
+
+
+class BaseEstimator(ABC):
+    """Estimator base class.
+
+    Base class for Estimator that estimates expectation values of quantum circuits and observables.
+    """
+
+    def __init__(
+        self,
+        circuits: Iterable[QuantumCircuit],
```

Comment:
Since there will be always be some capture at init time (here just capture to immutable tuple, but in some concrete implementation it can be serialized into session) I think it doesn't matter, since we would iterate over it at init time?

Reply:
Since we are using `len` in the code below this, it must be Sized. Or remove the len.

Reply:
Alternatively you can consume an `Iterable` into a tuple or list immediately on entry, since `tuple` is `Sized`, and that avoids any potential issues in accidentally mutating something that shouldn't be mutated.

Reply:
What @jakelishman said was what I had meant to write.